### ISSUE TITLE
Micropub Errors 

### ISSUE THREAD

**Issue by stephenpieper** [snarfed/wordpress-micropub#112]
I've started getting Micropub errors while using different services. I noticed it with OwnYourSwarm and Quill. The errors are the same in both cases. 

<b>Warning</b>:  array_diff(): Argument #1 is not an array in <b>/...wp-content/plugins/micropub/micropub.php</b> on line <b>252</b><br />

<b>Warning</b>:  Cannot modify header information - headers already sent by (output started at /...wp-content/plugins/micropub/micropub.php:252) in <b>.../wp-content/plugins/micropub/micropub.php</b> on line <b>1092</b><br />
{"error":"invalid_request","error_description":"Invalid Post Status"}

I haven't changed anything as far as I know.

I'm using Wordpress-Micropub v1.3 with other IndieWeb plugins like Simple Location.

**Comment by snarfed** [comment]
thanks for filing, and sorry for the trouble!

@dshanske, based on `Invalid Post Status`, mind helping debug?

**Comment by dshanske** [comment]
I have a solution, but I can't explain why the current code isn't working.

**Comment by snarfed** [comment]
looks like this is also #104.

**Comment by snarfed** [comment]
@stephenpieper we think this might be fixed now. if you're comfortable installing the current plugin code in github, feel free to try it and let us know if it works for you. thanks!

**Comment by stephenpieper** [comment]
I've installed the new code. The first error has gone but  but I'm still getting

{"error":"invalid_request","error_description":"Invalid Post Status"}

.

**Comment by dshanske** [comment]
I am still stumped. I am going to try completely redoing the code that generates the post status. I cannot reproduce this.

**Comment by stephenpieper** [comment]
I've installed the previous version of the plugin and it works so it's obviously something to do with 1.3.

**Comment by snarfed** [comment]
thank you for sticking with this, @stephenpieper, and for the debugging! that's valuable info.

i've merged @dshanske's latest fix attempt into master,  #115. let's see if that fixes it. if not, we can start narrowing down based on the diff between 1.2 and 1.3.

**Comment by stephenpieper** [comment]
No luck yet. I'm also getting this error with Quill if it's any use.

"Your endpoint did not return a Location header."

**Comment by dshanske** [comment]
I am going to keep working on this, but still stumped and working in the dark. All my manual tests and unit tests pass.

**Comment by dshanske** [comment]
We have two people reporting the problem, I wonder what the two of you have in common.

**Comment by dshanske** [comment]
What can you tell me about your installation?

**Comment by stephenpieper** [comment]
What information exactly would be useful to you? I think it's a pretty standard Wordpress installation. Wordpress 4.9.4 and the PHP version is 7.1.5.

**Comment by dshanske** [comment]
Honestly not sure. Looking for clues. I just checked with another user and his worked to specification. Stumped still.

**Comment by miklb** [comment]
Just spitballing here, but any chance @stephenpieper you are running any caching plugins?

**Comment by stephenpieper** [comment]
@miklb No. In fact I've tried turning off other plugins to see if any were causing problems but no change.

**Comment by aaronpk** [comment]
Can confirm that this problem does not exist on a clean wordpress install with the current 1.3 release. I'm trying to track down what causes the error to surface.

**Comment by miklb** [comment]
my experience with these types of bugs also leads me to inquire about php version, nginx vs apache, who the host is, etc.

**Comment by aaronpk** [comment]
Was able to reproduce this after carefully reading the other thread. The key is this: https://github.com/snarfed/wordpress-micropub/issues/104#issuecomment-359267767

If you save the writing settings page, it writes a blank value for `micropub_default_post_status` to the database, so then the `get_option` call later will return the blank value instead of the default.

**Comment by stephenpieper** [comment]
Installed the latest version and the errors are fixed. It's working for me again. Thanks.

**Comment by snarfed** [comment]
hey, that's great news! thanks for all the testing. and thanks to @dshanske for the fix!

**Comment by vishae** [comment]
I just attempted to post using Quill and I got the "Your endpoint did not return a Location header." error.

I'm using micropub 1.4.1, installed it from WP and not manual install though. That shouldn't make any difference right?

**Comment by dshanske** [comment]
It should not. What else can you tell us?

**Comment by vishae** [comment]
This is the raw response:

> HTTP/1.1 401 Unauthorized
> Server: nginx/1.12.2
> Date: Wed, 25 Apr 2018 12:23:49 GMT
> Content-Type: application/json
> Transfer-Encoding: chunked
> Connection: keep-alive
> Expires: Wed, 11 Jan 1984 05:00:00 GMT
> Cache-Control: no-cache, must-revalidate, max-age=0
> Link: <https://www.serenawho.com/wp-json/indieauth/1.0/token>; rel="token_endpoint", <https://www.serenawho.com/?micropub=endpoint>; rel="micropub", <https://www.serenawho.com/wp-json/webmention/1.0/endpoint>; rel="webmention", <https://www.serenawho.com/wp-json/webmention/1.0/endpoint>; rel="http://webmention.org/"
> X-Endurance-Cache-Level: 2
> 
> "Unauthorized"

I'm also getting the same response when trying to use OwnYourGram.

I'm not sure what other information you'd need. Do you want the [list of plugins](https://i1.wp.com/www.serenawho.com/wp-content/uploads/2018/04/2018-04-19_17-57-58.jpg?resize=768%2C1013&ssl=1) I have installed on WP? I haven't any custom codes on my WP install.

**Comment by dshanske** [comment]
@vishae Did you try logging out of Quill and logging back in? If you upgraded the IndieAuth plugin, then it would switch you to the new WordPress built-in IndieAuth system, which means your old Indieauth.com token won't work anymore.

**Comment by vishae** [comment]
@dshanske I've only ever logged into Quill using the latest IndieAuth plugin, so I don't think that's the case.

I did try logging out and logging back in again, but it's still giving me the same error.

**Comment by dshanske** [comment]
What were you trying to post in Quill?

**Comment by vishae** [comment]
@dshanske I should probably note that I tried using the micropub.rocks site to check my micropub and nothing goes through. They all give me the same error.

Is this something on my end that I need to fix?

**Comment by lostfocus** [comment]
I'm not too familiar with nginx, but could it be that it strips some headers before giving the request to WordPress? I had similar problems before.
